src/service/pom.rs

```diff
@@ -34,79 +37,145 @@ impl crate::server::MyService {
-                Ok(_) => {}
-                Err(err) => {
-                    return Err(Status::not_found(&format!(
-                        "protocol {} not found: {}",
-                        protocol_name, err
+            .join(protocol_name);
+        let lock = self.lock(&format!("_pom:{}", protocol_name)).await?;
+        let res = async {
+            let list_key = format!("protocol_operator_groups:{}", protocol_name);
+            let mut list = if self
+                ._internal_storage_contains(&request.get_ref().user_id, &list_key)
+                .await?
+            {
+                let list = self
+                    ._internal_storage_read(&request.get_ref().user_id, &list_key)
+                    .await?;
+                Message::decode(&*list).unwrap()
+            } else {
+                ListOfString { list: vec![] }
+            };
+            if request.get_ref().upgrade {
+                if list.list.is_empty() {
+                    match std::fs::remove_dir_all(&protocol_home) {
+                        Ok(_) => {}
+                        Err(err) => return Err(Status::internal(err.to_string())),
+                    }
+                } else {
+                    return Err(Status::aborted(format!(
+                        "Protocol {} has running instances.",
+                        protocol_name
                     )));
                 }
             }
-        }
-        let toml = match std::fs::read_to_string(&path).unwrap().parse::<Value>() {
-            Ok(toml) => toml,
-            Err(err) => return Err(Status::internal(err.to_string())),
-        };
-        if self.core_uri.is_none() {
-            return Err(Status::internal("core_uri not found."));
-        }
-        let core_addr = self.core_uri.as_ref().unwrap();
-        if toml.get("package").is_none() || toml["package"].get("entrypoint").is_none() {
-            return Err(Status::not_found("entrypoint not found."));
-        }
-        let entrypoint = toml["package"]["entrypoint"].as_str();
-        if entrypoint.is_none() {
-            return Err(Status::not_found("entrypoint not found."));
-        }
-        let entrypoint = entrypoint.unwrap();
-        let user_jwt = self
-            ._host_storage_read(&format!("users:{}:user_jwt", request.get_ref().user_id))
+            let toml_path = protocol_home.join("colink.toml");
+            if std::fs::metadata(&toml_path).is_err() {
+                match fetch_protocol_from_inventory(protocol_name, &colink_home).await {
+                    Ok(_) => {}
+                    Err(err) => {
+                        return Err(Status::not_found(&format!(
+                            "protocol {} not found: {}",
+                            protocol_name, err
+                        )));
+                    }
+                }
+            }
+            let toml = match std::fs::read_to_string(&toml_path)
```

Comment:
I know it's toml, but for what (in protocol package? the one in inventory)? Plz make sure the var name is meaningful, especially the func name is long and is not intuitive.